src/core/asyncUtils.ts

```diff
@@ -42,6 +63,107 @@ function asyncUtils(act: Act, addResolver: (callback: () => void) => void): Asyn
     return !timeoutSignal.timedOut
   }
 
+  /**
+   * `waitFor` implementation from `@testing-library/dom` for Jest fake timers
+   * @param callback
+   * @param param1
+   * @returns
+   */
+  const waitForInJestFakeTimers = (
+    callback: () => boolean | void,
+    {
+      interval,
+      stackTraceError,
+      timeout
+    }: { interval: number; timeout: number; stackTraceError: Error }
+  ) => {
+    return new Promise(async (resolve, reject) => {
+      let lastError: unknown
+      let finished = false
+
+      const overallTimeoutTimer = setTimeout(handleTimeout, timeout)
+
+      checkCallback()
+      // this is a dangerous rule to disable because it could lead to an
+      // infinite loop. However, eslint isn't smart enough to know that we're
+      // setting finished inside `onDone` which will be called when we're done
+      // waiting or when we've timed out.
+      // eslint-disable-next-line no-unmodified-loop-condition
+      while (!finished) {
+        if (!jestFakeTimersAreEnabled()) {
+          const error = new Error(
+            `Changed from using fake timers to real timers while using waitFor. This is not allowed and will result in very strange behavior. Please ensure you're awaiting all async things your test is doing before changing to real timers. For more info, please go to https://github.com/testing-library/dom-testing-library/issues/830`
+          )
+          copyStackTrace(error, stackTraceError)
+          reject(error)
+          return
+        }
+        // we *could* (maybe should?) use `advanceTimersToNextTimer` but it's
+        // possible that could make this loop go on forever if someone is using
+        // third party code that's setting up recursive timers so rapidly that
+        // the user's timer's don't get a chance to resolve. So we'll advance
+        // by an interval instead. (We have a test for this case).
+        jest.advanceTimersByTime(interval)
+
+        // It's really important that checkCallback is run *before* we flush
+        // in-flight promises. To be honest, I'm not sure why, and I can't quite
+        // think of a way to reproduce the problem in a test, but I spent
+        // an entire day banging my head against a wall on this.
+        checkCallback()
+
+        if (finished) {
+          break
+        }
+
+        // In this rare case, we *need* to wait for in-flight promises
+        // to resolve before continuing. We don't need to take advantage
+        // of parallelization so we're fine.
+        // https://stackoverflow.com/a/59243586/971592
+        await act(async () => {
```

Comment:
This is useless for React 17 and should be removed to not confuse. It is only needed in 18 where we need to remove the `act` from `waitFor`